test/dataframe.jl

```diff
@@ -1881,16 +1881,32 @@ end
     @test_throws ArgumentError push!(df, "a")
 end
 
-@testset "names for Type" begin
-    df = DataFrame(a1 = 1:3, a2 = [1, missing, 3],
-                   b1 = 1.0:3.0, b2 = [1.0, missing, 3.0],
-                   c1 = '1':'3', c2 = ['1', missing, '3'])
-    @test names(df, Int) == ["a1"]
-    @test names(df, Union{Missing, Int}) == ["a1", "a2"]
-    @test names(df, Real) == ["a1", "b1"]
-    @test names(df, Union{Missing, Real}) == ["a1", "a2", "b1", "b2"]
-    @test names(df, Any) == names(df)
-    @test names(df, Union{Char, Float64, Missing}) == ["b1", "b2", "c1", "c2"]
+@testset "names for Type, predicate + standard tests of cols" begin
+    df_long = DataFrame(a1 = 1:3, a2 = [1, missing, 3],
+                        b1 = 1.0:3.0, b2 = [1.0, missing, 3.0],
+                        c1 = '1':'3', c2 = ['1', missing, '3'], x=1:3)
+    for x in (df_long[:, Not(end)], @view(df_long[:, Not(end)]),
+              groupby(df_long[:, Not(end)], :a1), groupby(@view(df_long[:, Not(end)]), :a1),
+              eachrow(df_long[:, Not(end)]), eachrow(@view(df_long[:, Not(end)])),
+              eachcol(df_long[:, Not(end)]), eachcol(@view(df_long[:, Not(end)])),
+              df_long[1, Not(end)])
+        @test names(x, 1) == ["a1"]
+        @test names(x, "a1") == ["a1"]
+        @test names(x, :a1) == ["a1"]
+        @test names(x, [2, 1]) == ["a2", "a1"]
+        @test names(x, ["a2", "a1"]) == ["a2", "a1"]
+        @test names(x, [:a2, :a1]) == ["a2", "a1"]
+        @test names(x, Int) == ["a1"]
+        @test names(x, Int) == ["a1"]
+        @test names(x, Union{Missing, Int}) == ["a1", "a2"]
```

Comment:
As a user, when I ask for `names(x, Int)`, my intuition tells me that all `Int` columns should be considered regardless of whether it has missing values or not. Maybe it's better to pick a default behavior, and let the user choose the different behavior via another argument or a kwarg?

Reply:
@nalimilan - what do you think? I can see the usefulness of doing the union with `Missing` by default (most likely this is what you want typically) but then we need to add a kwarg to allow to opt-out of it.

Reply:
Well it's a difficult choice. Ideally we would be able to write `Int?` and everything would be nice. :-)

I agree that in most cases people probably don't care about whether the columns allow for missing values. But automatically assuming `Missing` makes things less obvious, so I don't know...

Reply:
I will go for what @tk3369 suggests and add `skipmissing` kwarg with default to `false`. The reason is that `Union{Missing, Int}` is not ideal, as it allows bare `Missing`. I will propose something in the update of the PR that I find reasonable.

Reply:
Done - the rule is complex, but I assume it reflects the intuitive expectation. I have listed all relevant cases in tests.

Reply:
Actually I think we should consider this in the context of https://github.com/JuliaData/DataFrames.jl/issues/2314. If we don't skip missing values by default anywhere, should we really assume `Union{T, Missing}` by default here? Maybe we should rather be strict and explicit here, but find a more concise syntax for DataFramesMeta so that by default it skips missing values and assumes `T` means `Union{T, Missing}`?

Reply:
Let us discuss it, but my point is that I do not implement `Union{T, Missing}` here. I implement something more careful, that is not easy to express using type system. The issue is that `Missing <: Union{T, Missing}` and we want to exclude this case, unless `T` is `Missing` (and some more corner cases like `Union{}`).

Anyway - there is no rush with merging this PR so we can think what is best.

Reply:
> The issue is that `Missing <: Union{T, Missing}` and we want to exclude this case, unless `T` is `Missing` (and some more corner cases like `Union{}`).

@nalimilan - this was my earlier assumption, but maybe we should change this as I described below.

The core of the question is if column contains only `missing` values and has `eltype` `Missing` should it match type e.g. `Int` but with `unionmissing=true`? (the thing is that this column does not allow the originaly passed type, e.g. `Int`)